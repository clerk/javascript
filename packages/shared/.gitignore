browser
callWithRetry
color
cookie
date
deprecated
error
file
globs
handleValueOrFn
isomorphicAtob
isomorphicBtoa
keys
loadScript
localStorageBroadcastChannel
poller
proxy
underscore
url
react
constants
apiUrlFromPublishableKey