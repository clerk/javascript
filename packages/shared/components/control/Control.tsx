import cn from 'classnames';
import React from 'react';

import { Label } from '../label';
// @ts-ignore
import styles from './Control.module.scss';

export type ControlProps = {
  children?: React.ReactNode;
  className?: string;
  containerStyle?: React.CSSProperties;
  error?: string;
  errorClassName?: string;
  errorStyle?: React.CSSProperties;
  footer?: React.ReactNode;
  hint?: React.ReactNode;
  hintOnClickHandler?: (e: React.MouseEvent) => void;
  htmlFor?: string;
  label?: React.ReactNode;
  labelClassName?: string;
  name?: string;
  optional?: boolean;
  sublabel?: React.ReactNode;
};

export function Control({
  children,
  className,
  containerStyle,
  error,
  errorClassName,
  errorStyle,
  footer,
  hint,
  hintOnClickHandler,
  htmlFor,
  label,
  labelClassName,
  optional,
  sublabel,
}: ControlProps): JSX.Element {
  // Inject hasError prop to children controls to render their error styles.
  // This convention should be enforced with a TS interface.
  const childrenWithErrorProp = React.Children.map(children, child => {
    if (React.isValidElement(child) && typeof child.type !== 'string') {
      return React.cloneElement(child, { hasError: !!error });
    }
    return child;
  });

  const hintElement = hint ? (
    <span className={styles.hint}>
      {hintOnClickHandler ? <a onClick={hintOnClickHandler}>{hint}</a> : hint}
    </span>
  ) : null;

  const optionalElement = optional ? (
    <span className={styles.optional}>Optional</span>
  ) : null;

  return (
    <div className={cn(styles.control, className)} style={containerStyle}>
      <Label htmlFor={htmlFor} className={styles.labelContainer}>
        {childrenWithErrorProp}
        {sublabel && <span className={cn(styles.sublabel)}>{sublabel}</span>}
        {label && (
          <span className={cn(styles.label, labelClassName)}>
            {label}
            {optionalElement}
            {hintElement}
          </span>
        )}
      </Label>
      <span
        className={cn(
          styles.error,
          { [styles.errorVisible]: error },
          errorClassName,
        )}
        style={errorStyle}
      >
        {error}
      </span>
      {footer}
    </div>
  );
}
