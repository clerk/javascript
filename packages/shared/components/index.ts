export * from './button';
export * from './card';
export * from './checkbox';
export * from './dropdown';
export * from './form';
export * from './input';
export * from './label';
export * from './modal';
export * from './outline';
export * from './popover';
export * from './portal';
export * from './radio';
export * from './spinner';
export * from './switch';
export * from './tabs';
export * from './titledCard';
export * from './toggle';
export * from './tooltip';
export * from './wizard';
export * from './heading';
export * from './oneTimeCodeInput';
export * from './tag';
export * from './textArea';
export * from './list';
export * from './avatar';
export * from './phoneInput';
export * from './control';
export * from './imageUploader';
