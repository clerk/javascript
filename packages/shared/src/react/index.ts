export * from './hooks';

export {
  ClerkInstanceContext,
  ClientContext,
  OptionsContext,
  OrganizationProvider,
  SessionContext,
  useAssertWrappedByClerkProvider,
  useClerkInstanceContext,
  useClientContext,
  useOptionsContext,
  useOrganizationContext,
  UserContext,
  useSessionContext,
  useUserContext,
} from './contexts';
