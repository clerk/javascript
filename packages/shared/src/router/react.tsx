/**
 * React-specific binding's for interacting with Clerk's router interface.
 */
import React, { createContext, useContext } from 'react';

import type { ClerkHostRouter, ClerkRouter } from './router';
import { createClerkRouter } from './router';

export const ClerkHostRouterContext = createContext<ClerkHostRouter | null>(null);
export const ClerkRouterContext = createContext<ClerkRouter | null>(null);

export function useClerkHostRouter() {
  const ctx = useContext(ClerkHostRouterContext);

  if (!ctx) {
    throw new Error(
      'clerk: Unable to locate ClerkHostRouter, make sure this is rendered within `<ClerkHostRouterContext.Provider>`.',
    );
  }

  return ctx;
}

export function useClerkRouter() {
  const ctx = useContext(ClerkRouterContext);

  if (!ctx) {
    throw new Error('clerk: Unable to locate ClerkRouter, make sure this is rendered within `<Router>`.');
  }

  return ctx;
}

/**
 * Construct a Clerk Router using the provided host router. The router instance is accessible using `useClerkRouter()`.
 */
export function Router({
  basePath,
  children,
  router,
}: {
  children: React.ReactNode;
  basePath?: string;
  router?: ClerkHostRouter;
}) {
  const hostRouter = useClerkHostRouter();
  const clerkRouter = createClerkRouter(router ?? hostRouter, basePath);

  return <ClerkRouterContext.Provider value={clerkRouter}>{children}</ClerkRouterContext.Provider>;
}

type RouteProps = { path?: string; index?: boolean };

/**
 * Used to conditionally render its children based on whether or not the current path matches the provided path.
 */
export function Route({ path, children, index }: RouteProps & { children: React.ReactNode }) {
  const parentRouter = useClerkRouter();

  if (!path && !index) {
    return children;
  }

  if (!parentRouter?.match(path, index)) {
    return null;
  }

  return children;
}
