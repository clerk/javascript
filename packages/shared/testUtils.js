module.exports = require('./utils/testUtils');
