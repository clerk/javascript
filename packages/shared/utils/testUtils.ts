import '@testing-library/jest-dom/extend-expect';

import {render, RenderOptions} from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';

import {noop} from '../utils';

// Wrap components with test providers
const customRender = (ui: React.ReactElement, options?: RenderOptions) => {
  return render(ui, options);
};

// To be used for structural snapshot testing
const renderJSON = (c: React.ReactElement) => renderer.create(c).toJSON();

// To be used for structural snapshot testing after the first rerender on mount
const renderJSONAfterFirstAct = (c: React.ReactElement) => {
  let root;

  renderer.act(() => {
    root = renderer.create(c);
  });

  // @ts-ignore
  return root.toJSON();
};

// Re-export everything
export * from '@testing-library/react';
export { act as actHook, renderHook } from '@testing-library/react-hooks';
export { default as userEvent } from '@testing-library/user-event';
export { mocked } from 'jest-mock';

// Override render method
export { customRender as render, renderJSON, renderJSONAfterFirstAct, noop };
