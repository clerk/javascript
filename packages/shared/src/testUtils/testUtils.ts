import '@testing-library/jest-dom/extend-expect';

import type { RenderOptions } from '@testing-library/react';
import { render } from '@testing-library/react';

import { noop } from '../utils';

// Wrap components with test providers
const customRender = (ui: React.ReactElement, options?: RenderOptions) => {
  return render(ui, options);
};

// Re-export everything
export * from '@testing-library/react';
export { default as userEvent } from '@testing-library/user-event';

// Override render method
export { customRender as render, noop };
