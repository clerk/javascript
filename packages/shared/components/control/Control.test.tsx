import { renderJSON } from '@clerk/shared/testUtils';
import * as React from 'react';

import { Control } from './Control';

describe('<Control/>', () => {
  it('renders the label along with its control, hints, optional status, and errors', () => {
    const tree = renderJSON(
      <Control
        label="foo"
        sublabel="bar"
        hint="qux"
        error="this is an error"
        errorStyle={{ display: 'block' }}
        optional
      >
        Foo
      </Control>
    );
    expect(tree).toMatchSnapshot();
  });
});
