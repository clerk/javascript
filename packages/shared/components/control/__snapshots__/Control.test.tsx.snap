// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Control/> renders the label along with its control, hints, optional status, and errors 1`] = `
<div
  className="control"
>
  <label
    className="label labelContainer"
  >
    Foo
    <span
      className="sublabel"
    >
      bar
    </span>
    <span
      className="label"
    >
      foo
      <span
        className="optional"
      >
        Optional
      </span>
      <span
        className="hint"
      >
        qux
      </span>
    </span>
  </label>
  <span
    className="error errorVisible"
    style={
      Object {
        "display": "block",
      }
    }
  >
    this is an error
  </span>
</div>
`;
